implement synchronization with blockchain via bitcoind's rpcclient

evgeniy-scherbina opened the pull request:
I plan to implement bitcoin wallet(something like GoLang's btcwallet) which can be backed by full-node(bitcoind for example) or spv(aka GoLang's neutrino). It's my initial effort.

So this implementation could be start of extending one more backend with bitcoind, in this case we will have two supported backends from out of the box: spv and bitcoind.

tamasblummer commented:
This PR as is lacks separation of generic wallet logic and specific bitcoind rpc implementation, eg. through interfaces. The rpc implementation itself pulls in a very long list of dependencies. Do not support to merge as is.

tamasblummer commented:
Yes, I am looking forward to it.

evgeniy-scherbina commented:
I temporary remove grpc, I'll plan to move it in separate crate(something like grpc-rust-wallet).
I rewrote communication with bitcoind to make bitcoin_rpc_client dependency optional.
I refactored communication with database.

I'll plan: 
1) move BlockChainIO interface impplementation in separate crase as well
2) add on-chain fee configuration
3) add integration with BIP39 implementation
4) add more tests, refactor code

evgeniy-scherbina commented:
What do you think about current state of PR?
Do you consider to accept this PR(as a base) and after that improve codebase, add features, add tests, etc.

tamasblummer commented:
I have a problem compiling this on Ubuntu 18.04.1 LTS with  gcc 7.3.0
as follows:
---------------------
   Compiling librocksdb-sys v5.14.2
error: failed to run custom build command for `librocksdb-sys v5.14.2`
process didn't exit successfully: `/home/tamas/Documents/rust-wallet/target/debug/build/librocksdb-sys-24328211cb65612c/build-script-build` (exit code: 101)
--- stdout
cargo:rerun-if-changed=build.rs
cargo:rerun-if-changed=rocksdb/
cargo:rerun-if-changed=snappy/

--- stderr
rocksdb/include/rocksdb/c.h:68:10: fatal error: 'stdarg.h' file not found
rocksdb/include/rocksdb/c.h:68:10: fatal error: 'stdarg.h' file not found, err: true
thread 'main' panicked at 'unable to generate rocksdb bindings: ()', libcore/result.rs:945:5

tamasblummer commented:
you might want to consider using hammersbald (https://github.com/rust-bitcoin/hammersbald) instead of rocksdb as it is smaller faster ant pure safe rust.

evgeniy-scherbina commented:
1) you can check circleci/config.yml file https://github.com/LightningPeach/rust-wallet/blob/split_lib_and_grpc/.circleci/config.yml#L22 to fix your issue
2) I've read about hammersbald on medium post when I learn rust-bitcoin-spv repo. Probably I replaced rocksdb with it, or maybe left both as plugins to rust-wallet.

tamasblummer commented:
Thanks, it compiles after linking those .h files. 
It seems to me that RocksDB use is append only except delete_utxo which is probably not even needed. Please make an attempt to implement these with Hammersbald also to help that project. I would like to reduce external dependencies esp. to C++ code.

tamasblummer commented:
I am ready to merge this, because this project is rather early in its development. I hope we can collaborate to reduce the number of dependencies and keep this useful also for projets that use bitcoin_spv instead of bitcoind

tamasblummer commented:
Before merge it would make sense to agree on following header for all files:
//
// Copyright 2018 rust-wallet developers
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

and add a NOTICE file with the following content:
rust-wallet
Copyright 2018 rust-wallet developers
This product includes software developed by
Tamas Blummer
(add your name or company)
